source/docs/h2_metadata.md

```diff
@@ -171,3 +169,37 @@ ConnectionManagerImpl::ActiveStream::encodeMetadata(ActiveStreamEncoderFilter\*
 to go through all the encoding filters.
 Or new metadata can be added to metadata\_map in
 StreamFilter::encodeMetadata(MetadataMap& metadata\_map) directly.
+
+## Request metadata handling
+
+We first explain how request metadata get consumed or proxied.
+In function EnvoyConnectionManagerImpl::ActiveStream::decodeMetadata(ActiveStreamDecoderFilter\* filter,
+MetadataMap& metadata\_map), Envoy passes request metadata received from downstream to filters by
+calling the following filter interface:
+
+FilterMetadatasStatus StreamDecoderFilter::decodeMetadata(MetadataMap& metadata\_map).
+
+Filters, by implementing the interface, can consume or modify request metadata. If no filter
+touches the metadata, it is proxied to upstream unchanged.
+
+The last filter in the filter chain is router filter. The router filter calls
+Filter::request\_encoder\_-\>encodeMetadata(const MetadataMapVector& metadata\_map\_vector) to pass
+the metadata to codec, and codec encodes and forwards the metadata to the upstream. If the connection
+to the upstream has not been established when metadata is received, the metadata is temporarily stored in
+Filter::downstream\_metadata\_map\_vector\_. When the connection is ready
+(Filter::UpstreamRequest::onPoolReady()), the metadata is then passed to codec, and forwarded to
+the upstream.
+
+Envoy can also add new request metadata through filters's decoding interfaces (See section
+[Inserting metadata](#inserting-metadata) for detailed interfaces). Filters can add new
+metadata to ActiveStream::request\_metadata\_map\_vector\_ by calling
+StreamDecoderFilterCallbacks::addDecodedMetadata(). After calling each filter's decoding function,
+Envoy checks if new metadata is added to ActiveStream::request\_metadata\_map\_vector\_. If so,
+then Envoy calls ConnectionManagerImpl::ActiveStream::decodeMetadata(ActiveStreamEncoderFilter\* filter,
+MetadataMapPtr&& metadata\_map) to go through all the filters.
```

Comment:
Should the newly added metadata go through ALL the filters or the current one and following filters? This PR seems to iterate through all the filters on the new metadata. But previous filters may already regards itself done with the stream (end_stream_ == true). So it's a bit surprising to me to know that the metadata added by decodeData|Trailers  should go through all the filters. I maybe missed some important point, if that's the case, please excuse my ignorance.

Reply:
Good question! You are right, metadata should start iterating from the current and the following filters. It was a decision made a few PRs back: (please refer to https://github.com/envoyproxy/envoy/pull/5165 for details and search for "current_encoder_filter_" for the discussion around this). About the end_stream concern, as long as end_stream is sent after metadata, we are OK, :) This is handled in router filter.